lib/api/apiUtils/quotas/quotaUtils.js

```diff
@@ -0,0 +1,231 @@
+const async = require('async');
+const { errors } = require('arsenal');
+const {
+    actionNeedQuotaCheckCopy,
+    actionNeedQuotaCheck,
+    actionWithDataDeletion,
+} = require('arsenal').policies;
+const { config } = require('../../../Config');
+const QuotaService = require('../../../quotas/quotas');
+
+/**
+ * Process the bytes to write based on the request and object metadata
+ * @param {string} apiMethod - api method
+ * @param {BucketInfo} bucket - bucket info
+ * @param {string} versionId - version id of the object
+ * @param {number} contentLength - content length of the object
+ * @param {object} objMD - object metadata
+ * @return {number} processed content length
+ */
+function processBytesToWrite(apiMethod, bucket, versionId, contentLength, objMD) {
+    let bytes = contentLength;
+    if (apiMethod === 'objectRestore') {
+        // object is being restored
+        bytes = Number.parseInt(objMD['content-length'], 10);
+    } else if (!bytes && objMD?.['content-length']) {
+        if (apiMethod === 'objectCopy') {
+            // object is being copied
+            bytes = Number.parseInt(objMD['content-length'], 10);
```

Comment:
the target of the copy may already exist, and should be accounted for in non-versioned buckets.

and we actually have this case in the product, when updating metadata (from the UI) in case bucket is not versioned...

Reply:
You are right, I updated the logic by passing an optional parameter: `destObjMD`, known at the quota evaluation time for Copy APIs, so that we can, in the `processBytesToWrite` function, know if the copy is a replacement. This is unit tested.
Also added a filter on the copy part because it was missing

Reply:
It's in https://github.com/scality/cloudserver/pull/5554/commits/3ca848057cff1ddf55f357c24cb8759aad1c2f94